File: src/main/java/au/id/micolous/metrodroid/card/classic/MfcCardImporter.java
package au.id.micolous.metrodroid.card.classic;

import android.support.annotation.NonNull;

import org.apache.commons.lang3.ArrayUtils;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.GregorianCalendar;
import java.util.Locale;

import au.id.micolous.metrodroid.card.CardImporter;
import au.id.micolous.metrodroid.key.ClassicSectorKey;
import au.id.micolous.metrodroid.xml.ImmutableByteArray;

public class MfcCardImporter implements CardImporter<ClassicCard> {
    private static final int MAX_SECTORS = 40;

    @NonNull
    @Override
    public ClassicCard readCard(@NonNull InputStream stream) throws IOException {
        // Read the blocks of the card.
        ArrayList<ClassicSector> sectors = new ArrayList<>();
        byte[] uid = null;
        int maxSector = 0;

    sectorloop:
        for (int sectorNum=0; sectorNum<MAX_SECTORS; sectorNum++) {
            ArrayList<ClassicBlock> blocks = new ArrayList<>();
            byte[] key = null;

            int blockCount = 4;
            if (sectorNum >= 32) {
                blockCount = 16;
            }

            for (int blockNum=0; blockNum<blockCount; blockNum++) {
                byte[] blockData = new byte[16];
                int r = stream.read(blockData);
                if (r <= 0 && blockNum == 0) {
                    // We got to the end of the file.
                    break sectorloop;
                } else if (r != blockData.length) {
                    throw new IOException(String.format(Locale.ENGLISH,
                            "Incomplete MFC read at sector %d block %d (%d bytes)",
                            sectorNum, blockNum, r));
                }

                if (sectorNum == 0 && blockNum == 0) {
                    // Manufacturer data
                    uid = ArrayUtils.subarray(blockData, 0, 4);
                    blocks.add(new ClassicBlock(blockNum, ClassicBlock.TYPE_MANUFACTURER, blockData));
                } else if (blockNum == blockCount - 1) {
                    key = ArrayUtils.subarray(blockData, 0, 6);
                    blocks.add(new ClassicBlock(blockNum, ClassicBlock.TYPE_TRAILER, blockData));
                } else {
                    blocks.add(new ClassicBlock(blockNum, ClassicBlock.TYPE_DATA, blockData));
                }
            }

            assert key != null;
            sectors.add(new ClassicSector(sectorNum, blocks.toArray(new ClassicBlock[0]),
                    ClassicSectorKey.Companion.fromDump(ImmutableByteArray.Companion.fromByteArray(key),
                            ClassicSectorKey.KeyType.A, "mfc-dump")));
            maxSector = sectorNum;
        }

        // End of file, now see how many blocks we get
        if (maxSector <= 15) {
            maxSector = 15; // 1K
        } else if (maxSector <= 31) {
            maxSector = 31; // 2K
        } else if (maxSector <= 39) {
            maxSector = 39; // 4K
        }

        // Fill missing sectors as "unauthorised".
        while (sectors.size() <= maxSector) {
            sectors.add(new UnauthorizedClassicSector(sectors.size()));
        }

        return new ClassicCard(uid,
                GregorianCalendar.getInstance(), sectors, false);
    }
}

File: src/commonMain/kotlin/au/id/micolous/metrodroid/serializers/classic/MfcCardImporter.kt
package au.id.micolous.metrodroid.serializers.classic

import au.id.micolous.metrodroid.card.Card
import au.id.micolous.metrodroid.card.classic.ClassicCard
import au.id.micolous.metrodroid.card.classic.ClassicSector
import au.id.micolous.metrodroid.card.classic.ClassicSectorRaw
import au.id.micolous.metrodroid.card.classic.UnauthorizedClassicSector
import au.id.micolous.metrodroid.serializers.CardImporter
import au.id.micolous.metrodroid.time.TimestampFull
import au.id.micolous.metrodroid.util.ByteArrayInput
import au.id.micolous.metrodroid.util.ImmutableByteArray
import au.id.micolous.metrodroid.util.Input
import au.id.micolous.metrodroid.util.toImmutable

class MfcCardImporter : CardImporter {
    fun readCard(bin: ByteArray): Card = readCard(stream=ByteArrayInput(bin))
    override fun readCard(stream: Input): Card {
        // Read the blocks of the card.
        val sectors = mutableListOf<ClassicSector>()
        var uid: ImmutableByteArray? = null
        var maxSector = 0

        for (sectorNum in 0 until MAX_SECTORS) {
            val blockCount =
                if (sectorNum >= 32)
                    16
                else
                    4

            val sectorData = stream.readBytes(16 * blockCount)
            if (sectorData.isEmpty() && sectorNum != 0) {
                 // We got to the end of the file.
                 break
            }

            if (sectorData.size != 16 * blockCount) {
                throw Exception("Incomplete MFC read at sector $sectorNum (${sectorData.size} bytes)")
            }

            if (sectorNum == 0) {
                // Manufacturer data
                uid = block0ToUid(sectorData.sliceArray(0..15).toImmutable())
            }

            val lbOffset = 16 * (blockCount - 1)
            val keyA = sectorData.sliceArray((lbOffset+0)..(lbOffset+5)).toImmutable()
            val keyB = sectorData.sliceArray((lbOffset+10)..(lbOffset+15)).toImmutable()
            val blocks =
                (0 until blockCount).map {
                    sectorData.sliceArray((it*16)..(it*16+15)).toImmutable()
                }

            val raw = ClassicSectorRaw(blocks, keyA,
                    keyB, false, null)
            sectors.add(ClassicSector.create(raw))
            maxSector = sectorNum
        }

        // End of file, now see how many blocks we get
        when {
            maxSector <= 15 -> maxSector = 15 // 1K
            maxSector <= 31 -> maxSector = 31 // 2K
            maxSector <= 39 -> maxSector = 39 // 4K
        }

        // Fill missing sectors as "unauthorised".
        while (sectors.size <= maxSector) {
            sectors.add(UnauthorizedClassicSector())
        }

        return Card(uid!!,
                TimestampFull.now(), mifareClassic = ClassicCard(sectors))
    }

    companion object {
        private const val MAX_SECTORS = 40

        fun block0ToUid(block0: ImmutableByteArray) =
                if (block0[0] == 4.toByte() && block0.byteArrayToInt(8, 2) in listOf(0x0400, 0x4400))
                    block0.copyOfRange(0, 7)
                else
                    block0.copyOfRange(0, 4)
    }
}
